File: desktop/soma/src/main/services/daemon-client/mappers.ts
import type {
	DaemonEvent as GrpcDaemonEvent,
	JoinRequest,
	PageRecord,
	Space,
	SpaceMember,
} from "@soma/proto/daemon/v1/daemon";

import type { DaemonStreamEvent, StoredJoinRequest, StoredPage, StoredSpace, StoredSpaceMember } from "./types";

export function fromPageRecord(page: PageRecord): StoredPage {
	return {
		spaceId: page.spaceId,
		pageId: page.pageId,
		title: page.title,
		parentPageIds: page.parentPageIds ?? [],
		createdAtMs: Number(page.createdAtMs ?? Date.now()),
		updatedAtMs: Number(page.updatedAtMs ?? Date.now()),
	};
}

export function fromSpace(space: Space): StoredSpace {
	return {
		spaceId: space.spaceId,
		displayName: space.displayName,
		ownerPeerId: space.ownerPeerId,
		createdAt: Number((space.createdAt as unknown) ?? Date.now()),
	};
}

export function fromSpaceMember(member: SpaceMember): StoredSpaceMember {
	return {
		spaceId: member.spaceId,
		peerId: member.peerId,
		role: member.role,
		expiresAt: Number(member.expiresAt ?? 0),
	};
}

export function fromJoinRequest(request: JoinRequest): StoredJoinRequest {
	return {
		requestId: request.requestId,
		spaceId: request.spaceId,
		subjectPeerId: request.subjectPeerId,
		displayName: request.displayName,
		deviceName: request.deviceName,
		requestedRole: request.requestedRole,
		createdAt: Number(request.createdAt ?? 0),
	};
}

export function mapDaemonEvent(event: GrpcDaemonEvent): DaemonStreamEvent | null {
	if (event.joinDecision) {
		return {
			kind: "join-decision",
			fromPeerId: event.joinDecision.fromPeerId,
			spaceId: event.joinDecision.decision?.spaceId?.value,
		};
	}
	if (event.joinSubmitted) {
		return {
			kind: "join-submitted",
			requestId: event.joinSubmitted.requestId,
			targetPeerId: event.joinSubmitted.targetPeerId,
		};
	}
	if (event.joinFailed) {
		return {
			kind: "join-failed",
			targetPeerId: event.joinFailed.targetPeerId,
			error: event.joinFailed.error,
		};
	}
	if (!event.documentBlobAdded) return null;
	return {
		kind: "document-blob-added",
		spaceId: event.documentBlobAdded.spaceId,
		docId: event.documentBlobAdded.docId,
		cid: event.documentBlobAdded.cid,
		mime: event.documentBlobAdded.mime,
		size: Number(event.documentBlobAdded.size ?? 0),
		name: event.documentBlobAdded.name,
	};
}

File: backend/bins/daemon/src/grpc/mappers.rs
use std::time::{SystemTime, UNIX_EPOCH};

use soma_proto_build::daemon;
use soma_storage::{blobs::BlobMetadata, pages::Page};

pub(super) fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis()
        .min(i64::MAX as u128) as i64
}

pub(super) fn to_page_record(page: Page) -> daemon::PageRecord {
    daemon::PageRecord {
        space_id: page.space_id,
        page_id: page.page_id,
        title: page.title,
        parent_page_ids: page.parent_page_ids,
        created_at_ms: page.created_at_ms,
        updated_at_ms: page.updated_at_ms,
    }
}

pub(super) fn to_blob_metadata(blob: BlobMetadata) -> daemon::BlobMetadata {
    daemon::BlobMetadata {
        space_id: blob.space_id,
        cid: blob.cid,
        size: blob.size.max(0) as u64,
        mime: blob.mime,
        name: blob.name,
        created_at_ms: blob.created_at_ms,
        last_seen_ms: blob.last_seen_ms,
    }
}

pub(super) fn map_space_record(space: crate::services::space::SpaceRecord) -> daemon::Space {
    daemon::Space {
        space_id: space.space_id,
        display_name: space.display_name.unwrap_or_default(),
        owner_peer_id: space.owner_peer_id.unwrap_or_default(),
        created_at: space.created_at,
    }
}
